multichain-testing/test/xcs-swap-anything/helpers.ts

```diff
@@ -0,0 +1,826 @@
+    } catch (error) {
+      return false;
+    }
+  };
+
+  const createPool = async (chainA: SwapParty, chainB: SwapParty) => {
+    const clientRegistry = osmosisClient.registry;
+    osmosis.gamm.poolmodels.balancer.v1beta1.load(clientRegistry);
+
+    const MsgCreateBalancerPool =
+      osmosis.gamm.poolmodels.balancer.v1beta1.MsgCreateBalancerPool;
+    const MsgCreateBalancerPoolResponse =
+      osmosis.gamm.poolmodels.balancer.v1beta1.MsgCreateBalancerPoolResponse;
+
+    const fee = {
+      amount: [{ denom: 'uosmo', amount: '10000' }],
+      gas: '1000000',
+    };
+
+    const inDenom = await getFinalDenom('osmosis', chainA);
+    const outDenom = await getFinalDenom('osmosis', chainB);
+
+    const message = MsgCreateBalancerPool.fromPartial({
+      sender: osmosisAddress,
+      poolParams: { swapFee: '0.01', exitFee: '0.00' },
+      poolAssets: [
+        { token: { denom: inDenom, amount: chainA.amount || '' }, weight: '1' },
+        {
+          token: {
+            denom: outDenom,
+            amount: chainB.amount || '',
+          },
+          weight: '1',
+        },
+      ],
+      futurePoolGovernor: '',
+    });
+    /** @type {Array<import('@cosmjs/proto-signing').EncodeObject>} */
+    const encodeObjects = [
+      {
+        typeUrl:
+          '/osmosis.gamm.poolmodels.balancer.v1beta1.MsgCreateBalancerPool',
+        value: message,
+      },
+    ];
+
+    const result = await osmosisClient.signAndBroadcast(
+      osmosisAddress,
+      encodeObjects,
+      fee,
+    );
+
+    const poolId = MsgCreateBalancerPoolResponse.decode(
+      result.msgResponses[0].value,
+    ).poolId;
+    return { poolId, inDenom, outDenom };
+  };
+
+  const setPoolRoute = async (
+    tokenIn: string,
+    tokenOut: string,
+    poolId: string,
+  ) => {
+    const swaprouterAddress = xcsContracts.swaprouter.address;
+    if (!swaprouterAddress) {
+      throw new Error('swaprouter contract address not found');
+    }
+
+    const txMsg = {
+      set_route: {
+        input_denom: tokenIn,
+        output_denom: tokenOut,
+        pool_route: [
+          {
+            pool_id: poolId,
+            token_out_denom: tokenOut,
+          },
+        ],
+      },
+    };
+
+    await invokeOsmosisContract(swaprouterAddress, txMsg);
+  };
+
+  const queryContractsInfo = async () => {
+    const osmosisCLI =
+      'kubectl exec -i osmosislocal-genesis-0 -c validator -- /bin/bash -c';
+
+    const info = `${osmosisCLI} "jq . /xcs-contracts-info.json"`;
+
+    const { stdout } = await execa(info, {
+      shell: true,
+    });
+
+    return JSON.parse(stdout);
+  };
+
+  /*
+   * Note: `getDenomHash` should only be called after the issuingDenom token has been successfully
+   * transferred from the issuingChain to currentChain (e.g., using fundRemote() or a similar mechanism).
+   * If no such transfer has occurred, the function will return 'undefined' because the denomination
+   * is not yet recognized on the currentChain.
```

Comment:
use `@returns undefined if ...`

Are the callers prepared to deal with `undefined`? If not, this should throw instead, right?

Reply:
Yes, the desired behaviour is for `getDenomHash` to throw if no hash found.
Change request implemented on commit: 9bdf4e3ca6d289d459fa0b7d514776e6dfe20231